src/api/app/models/worker_status.rb

```diff
@@ -4,33 +4,25 @@ def hidden
       mydata = Rails.cache.read('workerstatus')
       ws = Nokogiri::XML(mydata || Backend::Api::BuildResults::Worker.status).root
       # remove information about projects which are not visible to the current user
-      hide_project_information(ws)
+      hidden_projects(ws)
     end
 
     private
 
-    def hide_project_information(worker_status_root)
+    def hidden_projects(worker_status_root)
       worker_status = worker_status_root
       prjs = initialize_projects(worker_status)
-      xpath_string = generate_xpath(project_names(prjs.keys))
-      worker_status.xpath(xpath_string).each do |b|
-        b['project'] = '---'
-        b['repository'] = '---'
-        b['package'] = '---'
+      prj_names = project_names(prjs.keys)
+
+      worker_status.css('building').each do |b|
+        next if prj_names.include?(b['project'])
```

Comment:
CSS selector is much slower than plain  ruby, even though looking from straight Big O point of view,  this each + check if list contains element should be worse. https://gist.github.com/vpereira/18ff44783b18176926d3ecd2dfabc481